src/aiofiles/os.py

```diff
@@ -69,8 +70,31 @@ async def walk(top, topdown=True, onerror=None, followlinks=False):
     Wraps the `os.walk` function.
     """
 
-    for content in os.walk(
-        top, topdown=topdown, onerror=onerror, followlinks=followlinks
-    ):
-        yield content
-        await asyncio.sleep(0)
+    def _walk(q: asyncio.Queue):
+        nonlocal is_over
+        try:
+            for row in os.walk(
+                top, topdown=topdown, onerror=onerror, followlinks=followlinks
+            ):
+                q.put_nowait(row)
+        finally:
+            is_over = True
+
+    loop = asyncio.get_running_loop()
+    queue = asyncio.Queue()
+
+    walker = partial(_walk, q=queue)
+    loop.run_in_executor(None, walker)
+
+    is_over = False
+    while True:
+        if is_over and queue.empty():
```

Comment:
Producer may set `is_over` to True, but the queue may be not empty yet, that is why the loop must keep on.